### ISSUE TITLE
[bug] - Guzzle hangs forever after uploading a file

### ISSUE THREAD

**Issue by csimpi** [guzzle/guzzle#2594]
**Guzzle version(s) affected**: ^6.3
**PHP version**:  7.3.13
**cURL version**: 7.58.0

**Description**  
I'd like to upload a file to Mapbox server with Guzzle.
https://docs.mapbox.com/api/maps/#create-a-tileset-source

When the file is less then ~250MB I'm getting the response of the server about the file uploading was successful. Interestingly, with larger files Guzzle hangs forever, never retrieves with any response, never throws any error, just hanging without any response.

I can confirm that the file uploading is successful because I've added `progress` callback function which writes a file on the server and I can see that the `$uploadedBytes` is slowly increasing until `$uploadedBytes === $uploadTotal`.

Then nothing happens, just waiting forever.

With a smaller file, everything works perfectly.

If I run exactly the same request with curl from the command line with the same file, everything works as excepted, I'm getting the server's response about the file upload was successful.

**How to reproduce**  

```
            $response = $this->post('/tilesets/v1/sources/'.$username.'/'.$id,[],[
                'debug' => true,
                'multipart' => [
                    [
                        'name'     => 'file',
                        //'contents' => Storage::disk('geojson-files')->get($src_file),
                        'contents' => fopen($src_file, 'r'),
                        //'contents' => '@'.$src_file,
                        //'contents' => file_get_contents($src_file),
                        'filename' => 'file.geojson'
                    ]
                ],
                'progress' => function(
                    $downloadTotal,
                    $downloadedBytes,
                    $uploadTotal,
                    $uploadedBytes
                ) {
                   $fs = fopen('/tmp/progress','w');
                   if($uploadTotal && $uploadedBytes)
                    fputs($fs,$uploadedBytes.'/'.$uploadTotal);
                },
            ]);
```

The same request from the command line works:

```
curl -X POST "https://api.mapbox.com/tilesets/v1/sources/cveley/locations-src?access_token=TOKEN" -F file=@/tmp/locations-map.ld --header "Content-Type: multipart/form-data"
```

**Possible Solution**  

**Additional context**

**Comment by csimpi** [comment]
UPDATE:
Still wrong, and I think this is a bug...

I've added this:

```

                'on_stats' => function (\GuzzleHttp\TransferStats $stats) use ($statsfile) {
                    fputs($statsfile,json_encode(['uri'=>$stats->getEffectiveUri(),'time'=>$stats->getTransferTime(),'stats'=>$stats->getHandlerStats()]));
                    if ($stats->hasResponse()) {
                        fputs($statsfile,json_encode(['code'=>$stats->getResponse()->getStatusCode(),]));
                    } else {
                        fputs($statsfile,json_encode(['error'=>$stats->getHandlerErrorData()]));
                    }
                }
```

I'm getting the final server's response here in the `$stats->getHandlerStats()`, it's appearing in my logfile, but Guzzle->post() still hanging, the function never returns.

**Comment by csimpi** [comment]
UPDATE:
Still wrong, and I think this is a bug...

I've added this:

```

                'on_stats' => function (\GuzzleHttp\TransferStats $stats) use ($statsfile) {
                    fputs($statsfile,json_encode(['uri'=>$stats->getEffectiveUri(),'time'=>$stats->getTransferTime(),'stats'=>$stats->getHandlerStats()]));
                    if ($stats->hasResponse()) {
                        fputs($statsfile,json_encode(['code'=>$stats->getResponse()->getStatusCode(),]));
                    } else {
                        fputs($statsfile,json_encode(['error'=>$stats->getHandlerErrorData()]));
                    }
                }
```

I'm getting the final stats here in the `$stats->getHandlerStats()`, it's appearing in my logfile, but Guzzle->post() still hanging, the function never returns. 

```
"stats":{
"url":"https:\/\/api.mapbox.com\/tilesets\/v1\/sources\/cveley\/locations-src?access_token=********************",
"content_type":"application\/json; charset=utf -8",
"http_code":200,
"header_size":571,
"request_size":402,
"filetime":-1,
"ssl_verify_result":0,
"redirect_count":0,
"total_time":106.100176,
"namelookup_time":4.4e-5,
"connect_time":4.5e-5,
"pretransfer_time":0.000153,
"size_upload":261810838,
"size_download":109,
"speed_do wnload":1,
"speed_upload":2467585,
"download_content_length":109,
"upload_content_length":261810838,
"starttransfer_time":0.007919,
"redirect_time":0,
"redirect_url":"",
"primary_ip":"52.85.122.77",
"certinfo":[
],
"primary_port":443,
"local_ip":"10.0.2.15",
"local_port":5210 6
}
```

**Comment by csimpi** [comment]
UPDATE:
Still wrong, and I think this is a bug...

I've added this:

```

                'on_stats' => function (\GuzzleHttp\TransferStats $stats) use ($statsfile) {
                    fputs($statsfile,json_encode(['uri'=>$stats->getEffectiveUri(),'time'=>$stats->getTransferTime(),'stats'=>$stats->getHandlerStats()]));
                    if ($stats->hasResponse()) {
                        fputs($statsfile,json_encode(['code'=>$stats->getResponse()->getStatusCode(),]));
                    } else {
                        fputs($statsfile,json_encode(['error'=>$stats->getHandlerErrorData()]));
                    }
                }
```

I'm getting the final stats here in the `$stats->getHandlerStats()`, it's appearing in my logfile, but Guzzle->post() still hanging, the function never returns. 

```
"stats":{
"url":"https:\/\/api.mapbox.com\/tilesets\/v1\/sources\/cveley\/locations-src?access_token=********************",
"content_type":"application\/json; charset=utf -8",
"http_code":200,
"header_size":571,
"request_size":402,
"filetime":-1,
"ssl_verify_result":0,
"redirect_count":0,
"total_time":106.100176,
"namelookup_time":4.4e-5,
"connect_time":4.5e-5,
"pretransfer_time":0.000153,
"size_upload":261810838,
"size_download":109,
"speed_do wnload":1,
"speed_upload":2467585,
"download_content_length":109,
"upload_content_length":261810838,
"starttransfer_time":0.007919,
"redirect_time":0,
"redirect_url":"",
"primary_ip":"52.85.122.77",
"certinfo":[
],
"primary_port":443,
"local_ip":"10.0.2.15",
"local_port":5210 6
}
```

I've set a limit to create and upload a smaller file, the content of the stats is almost the same, but I've found some differences, some reason some redundant `space` character appearing randomly in the first response, for example: `utf -8` instead `utf-8`, or the local port is `5210 6` instead of `52106`

When I uploaded a smaller file the `space` characters are not appearing in the stats:

"stats":{
"url":"https:\/\/api.mapbox.com\/tilesets\/v1\/sources\/cveley\/locations-src?access_token=************************",
"content_type":"application\/json; charset=utf-8 ",
"http_code":200,
"header_size":546,
"request_size":377,
"filetime":-1,
"ssl_verify_result":0,
"redirect_count":0,
"total_time":1.532997,
"namelookup_time":2.2e-5,
"connect_time":2.3e-5,
"pretransfer_time":0.000134,
"size_upload":409106,
"size_download":103,
"speed_download":67,
"speed_upload":267040,
"download_content_length":103,
"upload_content_length":409106,
"starttransfer_time":0.000167,
"redirect_time":0,
"redirect_url":"",
"primary_ip":"52.85.122.77",
"certinfo":[
],
"primary_port":443,
"local_ip":"10.0.2.15",
"local_port":52166
}

**Comment by csimpi** [comment]
UPDATE 2:

Some reason Guzzle re-retrying the old never returned requests again and again.
I was checking my log files and noticed that the files are continuously changing and getting data from the older requests.

It seems when I start a guzzle request, since it never returns I close the page but it's still re-retrying... 
When I start another request, then one again and again, more and more record are appearing in the log files... I have to killall php-fpm processes to stop it.

**Comment by csimpi** [comment]
UPDATE 3:
If I change the main domain from local IP address to a virtualhost everything works well even the big files.

**Comment by sagikazarmark** [comment]
Can you try the following:

- use the StreamHandler
- profile the request (eg. with [Blackfire](https://blackfire.io/)

From your last message it seems you are trying to send something to `127.0.0.1`. Is that correct? php-curl is known to misbehave with loopback IP, I have no idea why though.

**Comment by csimpi** [comment]
@sagikazarmark 

The same happened with StreamHandler.

The complete environment on my side:
I have a webapp with Laravel backend and Angular frontend.

When I was testing I used `ng serve` opened the webapp from `192.168.0.10:4200`.
The backend is behind: `vagrant.mtp.travel`

This way I was getting the same error during bigger file upload.

When I compiled the angular project and used everything from `vagrant.mtp.travel` the issue disappeared. 

Good to know that, the Angular doesn't do any upload, just making a GET request to the backend (Laravel) and it creates and uploads the file to the mapbox servers.

So basically the frontend nothing has to do with Guzzle, so I still don't understand why I can't get the response when I upload a bigger file.
